### PR TITLE
DataPlan system - Base Rate Service

### PR DESCRIPTION
**Pull request by salman-codewar**
## Purpose changes
Issues: #346 

**Base Rate Services**

- Upload base rates service
- Get base rates service
- Get base rate service

```js
service RatesService {
    rpc GetBaseRates (GetBaseRatesRequest) returns (GetBaseRatesResponse) {}
    rpc GetBaseRate(GetBaseRateRequest) returns (GetBaseRateResponse){}
    rpc UploadBaseRates(UploadBaseRatesRequest) returns (UploadBaseRatesResponse){}
}
```

### UploadBaseRates Test
**Args:**
```js
[Require] FileURL : String
[Require] EffectiveAt : String
[Require] SimType : SIM_TYPE_ENUM
```

UploadBaseRates

### GetBaseRates Test
**Args:**
```js
[Require] Country : String
[Require] SimType : SIM_TYPE_ENUM
[Optional] Provider : String
[Optional] To : String
[Optional] From : String
```

GetRates

### GetBaseRate Test
**Args:**
```js
[Require] Id : Base_Rate_Id
```

GetRate

### CODE REVIEW

**Review by delvatt — APPROVED** [review]
Hi guys.  Thank you very much for this "small" gift. 
Have a few initial comments, but I think it will be better to go through them during our session today, as this MR is huge! :D 
For now I'm going to approve it and we'll be working incrementally in smaller MRs chunks  for improvements.

**Review by markhor — APPROVED** [review]

**Review by delvatt — CHANGES_REQUESTED** [review]

File: systems/data-plan/base-rate/pkg/db/rate_repo.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+func NewBaseRateRepo(db sql.Db) *baseRateRepo {
+	return &baseRateRepo{
+		Db: db,
+	}
+}
+
+func (u *baseRateRepo) GetBaseRate(rateId uint64) (*Rate, error) {
+	var rate *Rate
+	result := u.Db.GetGormDb().First(&rate, "Id=?", rateId)
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/db/rate_repo.go:25]: Mixing pointers and pointer on pointers.

File: systems/data-plan/base-rate/pkg/server/rate.go [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	}
+
+	if !utils.IsFutureDate(effectiveAt) {
+		logrus.Infof("Date you provided is not a valid future date.",
+			fileUrl, effectiveAt, simType)
+
+		return nil, status.Errorf(codes.InvalidArgument, "date you provided is not a valid future date %qs", effectiveAt)
+	}
+
+	data, err := utils.FetchData(fileUrl)
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/server/rate.go:84]: Guys, Thanks for making these changes. See how clean this is. Just easier & enjoyable to read!

File: systems/data-plan/base-rate/pkg/server/rate.go [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+	if err != nil {
+		logrus.Infof("Error fetching data: %v", err.Error())
+		return nil, status.Errorf(codes.Internal, err.Error())
+	}
+
+	res := utils.ParseToModel(data, effectiveAt, simType)
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/server/rate.go:90]: Is Parsing to Model **a never failing operation?**

File: systems/data-plan/base-rate/pkg/server/rate.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+
+	err = b.baseRateRepo.UploadBaseRates(res)
+
+	if err != nil {
+		logrus.Error("error inserting rates" + err.Error())
+		return nil, grpc.SqlErrorToGrpc(err, "rate")
+	}
+
+	rates, err := b.baseRateRepo.GetBaseRates("", "",effectiveAt,"")
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/server/rate.go:99]: Why are we getting back those rates?

File: systems/data-plan/base-rate/pkg/utils/utils.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+	return rates
+}
+
+func IsFutureDate(date string) bool {
+	t, err := time.Parse(time.RFC3339, date)
+	if err != nil {
+		return false
+	}
+	today := time.Now().UnixNano() / int64(time.Millisecond)
+	return today < t.UnixNano()/int64(time.Millisecond)
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/utils/utils.go:79]: Maybe it should be more clear and easy to use Go's standard package functions like `func (t Time) Before(u Time) bool` and `func (t Time) After(u Time) bool` for comparing dates & times.

**Review by delvatt — CHANGES_REQUESTED** [review]

File: systems/data-plan/base-rate/pkg/utils/utils.go (hunk 5 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/utils/utils.go:79]: Thanks that more easy. Using `time.Now().Before(t)`

File: systems/data-plan/base-rate/pkg/server/rate.go (hunk 4 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/server/rate.go:99]: Thanks, `gorm` returns the newly inserted rates on pointer reference. Removed this code.

File: systems/data-plan/base-rate/pkg/server/rate.go (hunk 3 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/server/rate.go:90]: Yea it just map `RawRates` struct to `db.Rate` struct. And if data is empty array it returns empty slice of `db.Rate`

File: systems/data-plan/base-rate/pkg/db/rate_repo.go (hunk 1 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/db/rate_repo.go:25]: Addressed

File: systems/data-plan/base-rate/pkg/validations/validations.go [hunk 6]
```diff
@@ -0,0 +1,41 @@
+package validationErrors
```
Comment (delvatt) [inline systems/data-plan/base-rate/pkg/validations/validations.go:1]: Also, sorry,  one last thing that I've failed to mention in our session: You need now to add a bunch of tests  for your repo functions, server rpcs and utils functions.

**Review by delvatt — CHANGES_REQUESTED** [review]

File: systems/data-plan/base-rate/pkg/server/rate.go (hunk 3 above)
Comment (delvatt) [reply systems/data-plan/base-rate/pkg/server/rate.go]: Ok.

File: systems/data-plan/base-rate/cmd/server/main.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+}
+
+// initConfig reads in config file, ENV variables, and flags if set.
+func initConfig() {
+	serviceConfig = &pkg.Config{
+		DB: &uconf.Database{
+			DbName: pkg.SystemName,
+		},
+	}
+	err := config.NewConfReader(pkg.SystemName).Read(serviceConfig)
```
[the remark this answers is not in the document]
Comment (Brackleycassinga) [reply systems/data-plan/base-rate/cmd/server/main.go:44]: Addressed

File: systems/data-plan/base-rate/cmd/server/main.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+	initConfig()
+	rateDb := initDb()
+	runGrpcServer(rateDb)
+}
+
+// initConfig reads in config file, ENV variables, and flags if set.
+func initConfig() {
+	serviceConfig = &pkg.Config{
+		DB: &uconf.Database{
+			DbName: pkg.SystemName,
```
[the remark this answers is not in the document]
Comment (Brackleycassinga) [reply systems/data-plan/base-rate/cmd/server/main.go:41]: Addressed

File: systems/data-plan/base-rate/pkg/validations/validations.go (hunk 6 above)
Comment (salman-codewar) [reply systems/data-plan/base-rate/pkg/validations/validations.go:1]: Yea, We're planning to do that in separate PR.

Comment (delvatt) [reply systems/data-plan/base-rate/pkg/validations/validations.go:1]: Okay, Got it!

**Review by delvatt — APPROVED** [review]
Good to go!

File: systems/data-plan/base-rate/pkg/server/rate.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+	"github.com/ukama/ukama/systems/data-plan/base-rate/pkg"
+	"github.com/ukama/ukama/systems/data-plan/base-rate/pkg/db"
+	"github.com/ukama/ukama/systems/data-plan/base-rate/pkg/utils"
+	validations "github.com/ukama/ukama/systems/data-plan/base-rate/pkg/validations"
+)
+
+type BaseRateServer struct {
+	baseRateRepo   db.BaseRateRepo
+	baseRoutingKey msgbus.RoutingKeyBuilder
+	nameGenerator  namegenerator.Generator
```
[the remark this answers is not in the document]
Comment (Brackleycassinga) [reply systems/data-plan/base-rate/pkg/server/rate.go]: We have removed this legacy code.

**Pull request merged**